### PR TITLE
gpu: Correct display transfer flipping edge case

### PR DESCRIPTION
**Pull request by GPUCode**
Fixes https://github.com/citra-emu/citra/issues/1542

Citra (master) | Citra (PR)
-|-
tales_broken | TALES OF THE ABYSS_07 09 23_03 32 38 32
pacman_broken | PAC-MAN PARTY 3D_07 09 23_03 32 47 852

It turns out that doing a display transfer with an input_width larger than the output_width and a vertical flip (alongside crop_input_lines) produces skewed output. Below I've attached a test homebrew that recreates this particular case. It configures a 256x512 framebuffer with a 240x400 viewport and does a display transfer to the screen. This required a modification to citro3d to accomplish, namely setting the second dim parameter [here](https://github.com/devkitPro/citro3d/blob/66a0594e5df08a656ea011466f7e7e761d12b241/source/framebuffer.c#L99) to **GX_BUFFER_DIM(240, 400)**. This problem is relevant because of the linked issue. These games perform a display transfer with these exact parameters and thus have wrong output. Decompilation efforts have shown that they actively try to workaround this hardware quirk by offseting the display transfer output address. Citra does not emulate this behaviour so the game's attempt to compensate breaks rendering. The output became black after the format reinterpretation framework as it was incorrectly being detected in ValidateByReinterpretation as able to skip valdation.

Άτιτλο

Fixing this problem is quite simple in fact, we simply offset the output address the opposite way as the games do in this specific case. Doing this we can emulate the broken rendering:

3DS | Citra (master) | Citra (PR)
-|-|-
2023-09-07_01-16-24 380_top |  | 

The test also confirms that the gap is not written by the display transfer, otherwise simply offseting the address would have not been entirely accurate. The source framebuffer is being cleared to a color so it looks like follows, yet the gap is still black.

Άτιτλο

What was found is that the gap contains parts of the last written data to the screen framebuffer instead:

<table>
<thead>
  <tr>
    <td align="center"><b>Flip = 0, Crop = 1<b></td>
    <td rowspan="2">➔</td>
    <td align="center"><b>Flip = 1, Crop = 1<b></td>
  </tr>
  <tr>
    <td></td>
    <td></td>
  </tr>
</thead>
</table>

<table>
<thead>
  <tr>
    <td align="center"><b>Flip = 1, Crop = 0<b></td>
    <td rowspan="2">➔</td>
    <td align="center"><b>Flip = 1, Crop = 1<b></td>
  </tr>
  <tr>
    <td></td>
    <td></td>
  </tr>
</thead>
</table>

Using this offset does emulate this behaviour as well. It is applied only when input > output, as it's a no-op when input = output and input < output is not practical and produces very garbage output. The input/output height also does not seem to matter.

[cubemap.zip](https://github.com/citra-emu/citra/files/12543374/cubemap.zip)

- - -
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/6952)

### CODE REVIEW

File: src/core/hw/gpu.cpp [hunk 1]
```diff
@@ -127,8 +127,8 @@ static void MemoryFill(const Regs::MemoryFillConfig& config) {
 }
 
 static void DisplayTransfer(const Regs::DisplayTransferConfig& config) {
-    const PAddr src_addr = config.GetPhysicalInputAddress();
-    const PAddr dst_addr = config.GetPhysicalOutputAddress();
+    PAddr src_addr = config.GetPhysicalInputAddress();
```
Comment (SachinVin) [inline src/core/hw/gpu.cpp:130]: Any reason to remove const here?

File: src/core/hw/gpu.cpp [hunk 2]
```diff
@@ -164,6 +164,14 @@ static void DisplayTransfer(const Regs::DisplayTransferConfig& config) {
     if (VideoCore::g_renderer->Rasterizer()->AccelerateDisplayTransfer(config))
         return;
 
+    // Using flip_vertically alongside crop_input_lines produces skewed output on hardware.
+    // We have to emulate this because some games rely on it existing to render correctly.
```
Comment (Dragios) [inline src/core/hw/gpu.cpp:168]:
```suggestion
    // We have to emulate this because some games rely on this behaviour to render correctly.
```

**Comment by Dragios** [comment]
I recommend utilizing HTML code instead of markdown for presenting this information, as it provides more flexibility and overcomes certain limitations.

<table>
<thead>
  <tr>
    <td align="center"><b>Flip = 0, Crop = 1<b></td>
    <td rowspan="2">➔</td>
    <td align="center"><b>Flip = 1, Crop = 1<b></td>
  </tr>
  <tr>
    <td></a></td>
    <td></td>
  </tr>
</thead>
</table>

I have prepared the HTML code for your convenience. You may copy this and modify accordingly:
```
<table>
<thead>
  <tr>
    <td align="center"><b>Flip = 0, Crop = 1<b></td>
    <td rowspan="2">➔</td>
    <td align="center"><b>Flip = 1, Crop = 1<b></td>
  </tr>
  <tr>
    <td><img src=https://user-images.githubusercontent.com/47210458/266169557-6f177ead-2805-432c-ba30-df9a56ec981b.png></a></td>
    <td><img src=https://user-images.githubusercontent.com/47210458/266171797-eb26e2b1-3aa1-4a69-bf3d-a4a79608b9bf.png></td>
  </tr>
</thead>
</table>
```

File: src/video_core/rasterizer_cache/rasterizer_cache.h [hunk 3]
```diff
@@ -272,6 +272,12 @@ bool RasterizerCache<T>::AccelerateDisplayTransfer(const GPU::Regs::DisplayTrans
     dst_params.pixel_format = PixelFormatFromGPUPixelFormat(config.output_format);
     dst_params.UpdateParams();
 
+    if (config.flip_vertically && config.crop_input_lines &&
```
Comment (FearlessTobi) [inline src/video_core/rasterizer_cache/rasterizer_cache.h:275]: We should probably add the comment here as well.

**Review by FearlessTobi — APPROVED** [review]

File: src/core/hw/gpu.cpp (hunk 1 above)
Comment (GPUCode) [reply src/core/hw/gpu.cpp:130]: No reason other than to make the variables look symmetric :p

**Comment by GPUCode** [comment]
Thanks @Dragios I have corrected the typo and edited the PR with your table, looks a lot cleaner!

**Review by Steveice10 — APPROVED** [review]

File: src/core/hw/gpu.cpp (hunk 1 above)
Comment (SachinVin) [reply src/core/hw/gpu.cpp]: Symmetry shouldn't be a reason to ignore good coding practices 😉

Comment (GPUCode) [reply src/core/hw/gpu.cpp]: ~~Me obssesing on how code looks way more than I should >_<~~

**Comment by Dragios** [comment]
Consider using [**Quoting code**](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-code) to better emphasize code block next time. Better for readability as well.
Example: `GX_BUFFER_DIM(240, 400)`

**Comment by fantesykikachu** [comment]
Just for future reference, this entirely fixed Tales of the Abyss for me (AMD GPU on windows). Before the software render gave graphical corruption like in the screen shots, and OpenGL just gave me a black screen (although the sound played fine); now both software and hardware look basically perfect.

**Pull request merged**